Disable rpc.enhancedClasses by default at runtime

niloc132 opened the pull request:
Logs warnings at compile time, indicating which classes need to be cleaned up to remove this feature.

Mitigation for #9709

niloc132 committed:
Disable rpc.enhancedClasses by default at runtime

Logs warnings at compile time, indicating which classes need to be
cleaned up to remove this feature.

Mitigation for #9709

niloc132 commented:
The good news is that all tests passed with this change https://github.com/niloc132/gwt/actions/runs/7267798597

The bad news is that this means that the `rpc.enhancedClasses` feature apparently has no tests - probably only bad news in retrospect, given that we are trying to turn it off.

niloc132 commented:
In light of https://github.com/gwtproject/gwt/issues/9709#issuecomment-1863677096, it might be wise to generalize these error messages slightly, and if the developer doesn't believe that JPA/JDO is in use, to just remove the annotations and see if that still works.

jnehlmeier commented on user/src/com/google/gwt/user/rebind/rpc/SerializableTypeOracleBuilder.java:
```diff
@@ -869,6 +869,10 @@ public SerializableTypeOracle build(TreeLogger logger) throws UnableToCompleteEx
       if (tic.maybeEnhanced()
           || (enhancedClasses != null && enhancedClasses.contains(type.getQualifiedSourceName()))) {
+        logger.log(TreeLogger.WARN, "The class " + type.getQualifiedSourceName() + " is both " +
+                "referenced from configuration as rpc.enhancedClasses and has JPA annotations. " +
```
The message should be `...has JPA/JDO annotations or is explicitly configured as enhanced class using configuration property ...`

jnehlmeier commented on user/src/com/google/gwt/user/server/rpc/RemoteServiceServlet.java:
```diff
@@ -94,6 +96,20 @@ static SerializationPolicy loadSerializationPolicy(HttpServlet servlet,
+                        "more detail. on the vulnerability that this presents.");
+              } else {
+                servlet.log("ERROR: Service uses enhanced classes, which are unsafe. Review " +
+                        "build logs to see where this can be fixed, or set " +
```
Little more explicit:

`Service de-/serializes enhanced JPA/JDO classes which is unsafe. Review build logs to see which classes are affected, ..`

jnehlmeier commented on user/src/com/google/gwt/user/server/rpc/RemoteServiceServlet.java:
```diff
@@ -94,6 +96,20 @@ static SerializationPolicy loadSerializationPolicy(HttpServlet servlet,
+            if (serializationPolicy.hasClientFields()) {
+              if (ENABLE_ENHANCED_CLASSES) {
+                servlet.log("WARNING: Enhanced JPA client fields are in use for this " +
+                        "application. See https://github.com/gwtproject/gwt/issues/9709 for " +
```
`Service de-/serializes enhanced JPA/JDO classes which is unsafe. See ...`

jnehlmeier commented on user/src/com/google/gwt/user/server/rpc/RemoteServiceServlet.java:
```diff
@@ -94,6 +96,20 @@ static SerializationPolicy loadSerializationPolicy(HttpServlet servlet,
+              if (ENABLE_ENHANCED_CLASSES) {
+                servlet.log("WARNING: Enhanced JPA client fields are in use for this " +
+                        "application. See https://github.com/gwtproject/gwt/issues/9709 for " +
+                        "more detail. on the vulnerability that this presents.");
```
`...for more detail on ...`

niloc132 committed:
Review feedback on message wording, fix whitespace char

jnehlmeier commented on user/src/com/google/gwt/user/rebind/rpc/SerializableTypeOracleBuilder.java:
```diff
@@ -869,6 +869,11 @@ public SerializableTypeOracle build(TreeLogger logger) throws UnableToCompleteEx
 
       if (tic.maybeEnhanced()
           || (enhancedClasses != null && enhancedClasses.contains(type.getQualifiedSourceName()))) {
+        logger.log(TreeLogger.WARN, "The class " + type.getQualifiedSourceName() + " has JPA " +
```
To be consistent JPA/JDO should be used here as well.

niloc132 committed:
Review feedback on compiler warnings

FrankHossfeld commented on user/src/com/google/gwt/user/server/rpc/RemoteServiceServlet.java:
```diff
@@ -94,6 +96,20 @@ static SerializationPolicy loadSerializationPolicy(HttpServlet servlet,
+              if (ENABLE_ENHANCED_CLASSES) {
+                servlet.log("WARNING: Service deserializes enhanced JPA/JDO classes, which is " +
+                        "unsafe. See https://github.com/gwtproject/gwt/issues/9709 for more " +
+                        "detail on the vulnerability that this presents.");
```
it should be: "details on the vulnerability"

jnehlmeier replied:
As German I tend to use `details` because that is how we speak natively but English also has "more detail on" if you want to emphasize that some content exists which provides a more granular version of what has already been said. So I think `for more detail on` is correct here as the previous sentence has given a very condensed summary. I think "more details on" would be more appropriate if the text has already given some details but not all.
But as I am not a native speaker I am fine with whatever @niloc132 choses :)

FrankHossfeld replied:
yeah, as German I would say 'details'. Checked also on Google both is ok. It is up to @niloc132 :-)

niloc132 replied:
Both "sound" correct to me - it appears that "for more detail" would be incorrect, but "for more details" or "for more detail on" are both acceptable? I'm going to leave this as is.

niloc132 commented:
Testing to confirm the new behavior:

1. Started with the DynaTable example, confirmed it started and worked correctly
2. Added `<extend-configuration-property name="rpc.enhancedClasses" value="com.google.gwt.sample.dynatable.client.Person" />` to the .gwt.xml file
3. Confirmed that the compiler logged a warning:
    ```
    [java]             Rebinding com.google.gwt.sample.dynatable.client.SchoolCalendarService
    [java]                Invoking generator com.google.gwt.user.rebind.rpc.ServiceInterfaceProxyGenerator
    [java]                   Generating client proxy for remote service interface 'com.google.gwt.sample.dynatable.client.SchoolCalendarService'
    [java]                      [WARN] The class com.google.gwt.sample.dynatable.client.Person has JPA/JDO annotations or is explicitly configured as an enhanced class using the configuration property rpc.enhancedClasses. This makes the server vulnerable to an issue with deserialization of unsafe data. See https://github.com/gwtproject/gwt/issues/9709 for more information.
    ```
4. Confirmed that the server would not load the policy and logged a server error:
    ```
    [java] 2023-12-23 12:24:51.073:INFO:oejshC.ROOT:qtp559361838-84: calendar: ERROR: Service deserializes enhanced JPA/JDO classes, which is unsafe. Review build logs to see which classes are affected, or set gwt.enhancedClasses.enabled to true to allow using this service. See https://github.com/gwtproject/gwt/issues/9709 for more detail.
    [java] 2023-12-23 12:24:51.074:INFO:oejshC.ROOT:qtp559361838-84: calendar: WARNING: Failed to get the SerializationPolicy 'D08326A5FC5842E352DF103522C617ED' for module 'http://127.0.0.1:8888/dynatable/'; a legacy, 1.3.3 compatible, serialization policy will be used.  You may experience SerializationExceptions as a result.
    ```
5. Note also that the client could not load data correctly with the legacy policy, and had its own error.
6. Added `-Dgwt.enhancedClasses.enabled=true` to the JVM args, confirmed that the above error was replaced by a warning and the application worked:
    ```
    [java] 2023-12-23 12:38:39.032:INFO:oejshC.ROOT:qtp448000133-82: calendar: WARNING: Service deserializes enhanced JPA/JDO classes, which is unsafe. See https://github.com/gwtproject/gwt/issues/9709 for more detail on the vulnerability that this presents.
    ```
7. Next, removed the configuration property, and instead added `@javax.persistence.Entity` to the Person bean, `org.datanucleus:javax.persistence:2.2.3` added to `war/WEB-INF/lib/`, and repeated the test from step 3.
